src/server/scripts/Outland/CoilfangReservoir/TheSlavePens/boss_ahune.cpp

```diff
@@ -0,0 +1,1053 @@
+                me->Kill(frozenCore);
+
+            Map::PlayerList const& players = me->GetMap()->GetPlayers();
+            if (!players.isEmpty())
+            {
+                if (Group* group = players.begin()->GetSource()->GetGroup())
+                    if (group->isLFGGroup())
+                        sLFGMgr->FinishDungeon(group->GetGUID(), 286);
+            }
+        }
+
+        void JustSummoned(Creature* summon) override
+        {
+            BossAI::JustSummoned(summon);
+        }
+
+        void DoAction(int32 action) override
+        {
+            if (action == ACTION_AHUNE_RETREAT)
+            {
+                Submerge();
+                events.ScheduleEvent(EVENT_EMERGE, 35000);
+            }
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (!UpdateVictim())
+                return;
+
+            events.Update(diff);
+
+            while (uint32 eventId = events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                    case EVENT_INITIAL_EMERGE:
+                        DoCast(me, SPELL_STAND);
+                        DoCast(me, SPELL_AHUNE_SPANKY_HANDS);
+                        DoCast(me, SPELL_AHUNES_SHIELD);
+                        break;
+                    case EVENT_EMERGE:
+                        Emerge();
+                        break;
+                    case EVENT_SYNCH_HEALTH:
+                        if (Creature* frozenCore = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FROZEN_CORE)))
+                            DoCast(frozenCore, SPELL_SYNCH_HEALTH);
+                        else
+                            DoCast(me, SPELL_SUICIDE);
+                        events.ScheduleEvent(EVENT_SYNCH_HEALTH, 3000);
+                        break;
+                    default:
+                        break;
+                }
+            }
+            DoMeleeAttackIfReady();
+        }
+
+        void Emerge()
+        {
+            if (Creature* frozenCore = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FROZEN_CORE)))
+                frozenCore->AI()->DoAction(ACTION_AHUNE_RESURFACE);
+
+            DoCast(me, SPELL_AHUNES_SHIELD);
+            me->RemoveAurasDueToSpell(SPELL_AHUNE_SELF_STUN);
+            me->RemoveAurasDueToSpell(SPELL_STAY_SUBMERGED);
+            DoCast(me, SPELL_STAND);
+            DoCast(me, SPELL_RESURFACE, true);
+            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_NOT_SELECTABLE);
+            events.ScheduleEvent(EVENT_SYNCH_HEALTH, 3000);
+        }
+
+        void Submerge()
+        {
+            if (Creature* frozenCore = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FROZEN_CORE)))
+                frozenCore->AI()->DoAction(ACTION_AHUNE_RETREAT);
+            me->RemoveAurasDueToSpell(SPELL_AHUNES_SHIELD);
+            me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_NOT_SELECTABLE);
```

Comment:
Might want to add `UNIT_FLAG_IMMUNE_TO_PC | UNIT_FLAG_IMMUNE_TO_NPC` if you don't want him to take damage while submerged.